File: remi-contracts/src/git.rs
//! Git 操作模式定义。

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Git 状态查询的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStatusInput {
    /// 仓库路径。
    pub repo_path: String,
}

/// Git 状态查询的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStatusResult {
    /// 当前分支名称。
    pub current_branch: Option<String>,
    /// 仓库是否处于干净状态。
    pub is_clean: bool,
    /// 已暂存的变更。
    pub staged: Vec<GitFileChange>,
    /// 未暂存的变更。
    pub unstaged: Vec<GitFileChange>,
    /// 未跟踪的文件。
    pub untracked: Vec<String>,
}

/// Git 文件变更。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitFileChange {
    /// 文件路径。
    pub path: String,
    /// 变更状态。
    pub status: GitChangeStatus,
}

/// Git 变更状态。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
#[serde(rename_all = "UPPERCASE")]
pub enum GitChangeStatus {
    /// 新增。
    Added,
    /// 修改。
    Modified,
    /// 删除。
    Deleted,
    /// 重命名。
    Renamed,
    /// 复制。
    Copied,
    /// 未合并。
    Unmerged,
}

/// Git 切换操作的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCheckoutInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 要切换的分支或提交。
    pub target: String,
    /// 是否创建新分支。
    #[serde(default)]
    pub create_branch: bool,
}

/// 创建分支的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateBranchInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 分支名称。
    pub branch_name: String,
    /// 基准分支或提交（可选）。
    pub base: Option<String>,
    /// 是否立即切换到新分支。
    #[serde(default)]
    pub checkout: bool,
}

/// 创建分支的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateBranchResult {
    /// 分支名称。
    pub branch_name: String,
    /// 提交 SHA。
    pub commit_sha: String,
}

/// 列出分支的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitListBranchesInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 是否包含远程分支。
    #[serde(default)]
    pub include_remote: bool,
}

/// 列出分支的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitListBranchesResult {
    /// 分支列表。
    pub branches: Vec<GitBranch>,
}

/// Git 分支。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitBranch {
    /// 分支名称。
    pub name: String,
    /// 是否为当前分支。
    pub is_current: bool,
    /// 是否为远程分支。
    pub is_remote: bool,
    /// 提交 SHA。
    pub commit_sha: String,
}

/// Git 拉取操作的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitPullInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 远程名称（默认: "origin"）。
    pub remote: Option<String>,
    /// 分支名称（可选）。
    pub branch: Option<String>,
}

/// Git 拉取操作的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitPullResult {
    /// 拉取的提交数量。
    pub commits_pulled: u32,
    /// 是否存在冲突。
    pub has_conflicts: bool,
}

/// 读取工作树差异的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitReadWorkingTreeDiffInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 文件路径（可选，用于指定特定文件）。
    pub file_path: Option<String>,
    /// 是否包含已暂存的变更。
    #[serde(default)]
    pub include_staged: bool,
}

/// 读取工作树差异的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitReadWorkingTreeDiffResult {
    /// 差异内容（unified diff 格式）。
    pub diff: String,
}

/// Git 操作进度事件。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitActionProgressEvent {
    /// 操作 ID。
    pub operation_id: Uuid,
    /// 进度百分比（0-100）。
    pub progress: u8,
    /// 状态消息。
    pub message: String,
}

/// 创建工作树的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateWorktreeInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 工作树路径。
    pub worktree_path: String,
    /// 分支名称。
    pub branch_name: String,
    /// 是否创建分离的工作树。
    #[serde(default)]
    pub detached: bool,
}

/// 创建工作树的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateWorktreeResult {
    /// 工作树路径。
    pub worktree_path: String,
    /// 分支名称。
    pub branch_name: String,
}

/// 创建分离工作树的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateDetachedWorktreeInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 工作树路径。
    pub worktree_path: String,
    /// 提交 SHA。
    pub commit_sha: String,
}

/// 创建分离工作树的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitCreateDetachedWorktreeResult {
    /// 工作树路径。
    pub worktree_path: String,
}

/// 移除工作树的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitRemoveWorktreeInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 工作树路径。
    pub worktree_path: String,
    /// 是否强制移除。
    #[serde(default)]
    pub force: bool,
}

/// 初始化 Git 仓库的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitInitInput {
    /// 目录路径。
    pub path: String,
}

/// 差异摘要的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitSummarizeDiffInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 差异内容。
    pub diff: String,
}

/// 差异摘要的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitSummarizeDiffResult {
    /// 摘要文本。
    pub summary: String,
    /// 变更的文件数量。
    pub files_changed: u32,
    /// 新增行数。
    pub insertions: u32,
    /// 删除行数。
    pub deletions: u32,
}

/// 移除索引锁的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitRemoveIndexLockInput {
    /// 仓库路径。
    pub repo_path: String,
}

/// 暂存操作的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStashInfoInput {
    /// 仓库路径。
    pub repo_path: String,
}

/// 暂存信息的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStashInfoResult {
    /// 暂存条目列表。
    pub stashes: Vec<GitStashEntry>,
}

/// Git 暂存条目。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStashEntry {
    /// 暂存索引。
    pub index: u32,
    /// 暂存消息。
    pub message: String,
    /// 时间戳（ISO 8601 格式）。
    pub timestamp: String,
}

/// 删除暂存的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStashDropInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 暂存索引。
    pub index: u32,
}

/// 暂存并切换分支的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitStashAndCheckoutInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 要切换的分支。
    pub branch: String,
    /// 暂存消息（可选）。
    pub message: Option<String>,
}

/// 运行堆叠操作的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitRunStackedActionInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 操作类型。
    pub action: String,
    /// 操作参数。
    pub params: serde_json::Value,
}

/// 准备拉取请求的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitPreparePullRequestThreadInput {
    /// 仓库路径。
    pub repo_path: String,
    /// 基础分支。
    pub base_branch: String,
    /// 头部 分支。
    pub head_branch: String,
    /// PR 标题。
    pub title: String,
    /// PR 描述。
    pub description: Option<String>,
}

/// 准备拉取请求的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitPreparePullRequestThreadResult {
    /// PR 编号。
    pub pr_number: u32,
    /// PR URL。
    pub pr_url: String,
}

/// 解析拉取请求的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitResolvePullRequestResult {
    /// 仓库路径。
    pub repo_path: String,
    /// PR 编号。
    pub pr_number: u32,
}

/// 拉取请求引用查询的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitPullRequestRefInput {
    /// 仓库路径。
    pub repo_path: String,
    /// PR 编号。
    pub pr_number: u32,
}

/// 移交线程的输入参数。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitHandoffThreadInput {
    /// 线程 ID。
    pub thread_id: uuid::Uuid,
    /// 目标工作树路径。
    pub worktree_path: String,
}

/// 移交线程的结果。
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct GitHandoffThreadResult {
    /// 新线程 ID。
    pub new_thread_id: uuid::Uuid,
}

File: remi-app/src/contracts/git.ts
/**
 * @file Git 操作契约模块
 *
 * 本模块定义了 Remi 系统中与 Git 版本控制交互的所有 Schema，涵盖仓库状态查询、
 * 分支管理、提交、推送、Pull Request、Worktree 等操作。
 *
 * ## 核心契约
 *
 * - `GitStatusInput/Result`：仓库状态（分支、暂存、修改、未跟踪文件等）
 * - `GitBranchesInput/Result`：分支列表查询
 * - `GitCommitInput/Result`：提交操作
 * - `GitPushInput/Result`：推送操作
 * - `GitPullRequestInput/Result`：PR 创建（通过 GitHub CLI）
 * - `GitWorktreeInput/Result`：Worktree 创建/删除
 * - `GitTextGenerationInput/Result`：AI 生成提交信息 / PR 描述
 * - `GitHubAuthStatusResult`：GitHub 认证状态
 *
 * ## 协议流程
 *
 * 1. `gitStatus` 实时获取仓库状态，前端订阅状态变更事件
 * 2. AI 生成提交信息：`gitTextGeneration` → 人工确认 → `gitCommit`
 * 3. 推送并创建 PR：`gitPush` → `gitCreatePullRequest`
 * 4. Worktree 切换：`gitWorktreeCreate` → 在新 worktree 中继续工作
 *
 * ## 性能注意
 *
 * - `gitStatus` 在大型仓库中可能耗时较长，建议使用 `GitStatusBroadcaster` 缓存
 * - 提交信息生成应异步执行，避免阻塞 UI
 * - PR 创建依赖 `gh` CLI，需先检查 `GitHubAuthStatus`
 */

import { Option, Schema } from "effect";
import { NonNegativeInt, PositiveInt, TrimmedNonEmptyString } from "./baseSchemas";
import { DEFAULT_GIT_TEXT_GENERATION_MODEL } from "./model";
import { ProviderStartOptions } from "./orchestration";

const TrimmedNonEmptyStringSchema = TrimmedNonEmptyString;

// Domain Types

export const GitStackedAction = Schema.Literals([
  "commit",
  "push",
  "create_pr",
  "commit_push",
  "commit_push_pr",
]);
export type GitStackedAction = typeof GitStackedAction.Type;
export const GitActionProgressPhase = Schema.Literals(["branch", "commit", "push", "pr"]);
export type GitActionProgressPhase = typeof GitActionProgressPhase.Type;
export const GitActionProgressKind = Schema.Literals([
  "action_started",
  "phase_started",
  "hook_started",
  "hook_output",
  "hook_finished",
  "action_finished",
  "action_failed",
]);
export type GitActionProgressKind = typeof GitActionProgressKind.Type;
export const GitActionProgressStream = Schema.Literals(["stdout", "stderr"]);
export type GitActionProgressStream = typeof GitActionProgressStream.Type;
const GitCommitStepStatus = Schema.Literals([
  "created",
  "skipped_no_changes",
  "skipped_not_requested",
]);
const GitPushStepStatus = Schema.Literals([
  "pushed",
  "skipped_not_requested",
  "skipped_up_to_date",
]);
const GitBranchStepStatus = Schema.Literals(["created", "skipped_not_requested"]);
const GitPrStepStatus = Schema.Literals(["created", "opened_existing", "skipped_not_requested"]);
const GitStatusPrState = Schema.Literals(["open", "closed", "merged"]);
const GitPullRequestReference = TrimmedNonEmptyStringSchema;
const GitPullRequestState = Schema.Literals(["open", "closed", "merged"]);
const GitPreparePullRequestThreadMode = Schema.Literals(["local", "worktree"]);
const GitHandoffThreadMode = Schema.Literals(["local", "worktree"]);

export const GitBranch = Schema.Struct({
  name: TrimmedNonEmptyStringSchema,
  isRemote: Schema.optional(Schema.Boolean),
  remoteName: Schema.optional(TrimmedNonEmptyStringSchema),
  current: Schema.Boolean,
  isDefault: Schema.Boolean,
  worktreePath: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
});
export type GitBranch = typeof GitBranch.Type;

const GitWorktree = Schema.Struct({
  path: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema,
});
const GitDetachedWorktree = Schema.Struct({
  path: TrimmedNonEmptyStringSchema,
  ref: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
});
const GitResolvedPullRequest = Schema.Struct({
  number: PositiveInt,
  title: TrimmedNonEmptyStringSchema,
  url: Schema.String,
  baseBranch: TrimmedNonEmptyStringSchema,
  headBranch: TrimmedNonEmptyStringSchema,
  state: GitPullRequestState,
});
export type GitResolvedPullRequest = typeof GitResolvedPullRequest.Type;

// RPC Inputs

export const GitStatusInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitStatusInput = typeof GitStatusInput.Type;

export const GitReadWorkingTreeDiffInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  scope: Schema.optional(Schema.Literals(["workingTree", "unstaged", "staged", "branch"])).pipe(
    Schema.withConstructorDefault(() => Option.some("workingTree" as const)),
  ),
});
export type GitReadWorkingTreeDiffInput = typeof GitReadWorkingTreeDiffInput.Type;

export const GitPullInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitPullInput = typeof GitPullInput.Type;

// Read-only diff summary requests reuse the shared git text-generation model settings.
export const GitSummarizeDiffInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  patch: Schema.String,
  codexHomePath: Schema.optional(TrimmedNonEmptyStringSchema),
  providerOptions: Schema.optional(ProviderStartOptions),
  textGenerationModel: Schema.optional(TrimmedNonEmptyStringSchema).pipe(
    Schema.withConstructorDefault(() => Option.some(DEFAULT_GIT_TEXT_GENERATION_MODEL)),
  ),
});
export type GitSummarizeDiffInput = typeof GitSummarizeDiffInput.Type;

export const GitRunStackedActionInput = Schema.Struct({
  actionId: TrimmedNonEmptyStringSchema,
  cwd: TrimmedNonEmptyStringSchema,
  action: GitStackedAction,
  commitMessage: Schema.optional(TrimmedNonEmptyStringSchema.check(Schema.isMaxLength(10_000))),
  featureBranch: Schema.optional(Schema.Boolean),
  filePaths: Schema.optional(
    Schema.Array(TrimmedNonEmptyStringSchema).check(Schema.isMinLength(1)),
  ),
  codexHomePath: Schema.optional(TrimmedNonEmptyStringSchema),
  providerOptions: Schema.optional(ProviderStartOptions),
  textGenerationModel: Schema.optional(TrimmedNonEmptyStringSchema).pipe(
    Schema.withConstructorDefault(() => Option.some(DEFAULT_GIT_TEXT_GENERATION_MODEL)),
  ),
});
export type GitRunStackedActionInput = typeof GitRunStackedActionInput.Type;

export const GitListBranchesInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitListBranchesInput = typeof GitListBranchesInput.Type;

export const GitCreateWorktreeInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema,
  newBranch: Schema.optional(TrimmedNonEmptyStringSchema),
  path: Schema.NullOr(TrimmedNonEmptyStringSchema),
});
export type GitCreateWorktreeInput = typeof GitCreateWorktreeInput.Type;

export const GitCreateDetachedWorktreeInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  ref: TrimmedNonEmptyStringSchema,
  path: Schema.NullOr(TrimmedNonEmptyStringSchema),
});
export type GitCreateDetachedWorktreeInput = typeof GitCreateDetachedWorktreeInput.Type;

export const GitPullRequestRefInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  reference: GitPullRequestReference,
});
export type GitPullRequestRefInput = typeof GitPullRequestRefInput.Type;

export const GitPreparePullRequestThreadInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  reference: GitPullRequestReference,
  mode: GitPreparePullRequestThreadMode,
});
export type GitPreparePullRequestThreadInput = typeof GitPreparePullRequestThreadInput.Type;

export const GitHandoffThreadInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  targetMode: GitHandoffThreadMode,
  currentBranch: Schema.NullOr(TrimmedNonEmptyStringSchema),
  worktreePath: Schema.NullOr(TrimmedNonEmptyStringSchema),
  associatedWorktreePath: Schema.NullOr(TrimmedNonEmptyStringSchema),
  associatedWorktreeBranch: Schema.NullOr(TrimmedNonEmptyStringSchema),
  associatedWorktreeRef: Schema.NullOr(TrimmedNonEmptyStringSchema),
  preferredLocalBranch: Schema.NullOr(TrimmedNonEmptyStringSchema),
  preferredWorktreeBaseBranch: Schema.NullOr(TrimmedNonEmptyStringSchema),
  preferredNewWorktreeName: Schema.NullOr(TrimmedNonEmptyStringSchema),
});
export type GitHandoffThreadInput = typeof GitHandoffThreadInput.Type;

export const GitRemoveWorktreeInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  path: TrimmedNonEmptyStringSchema,
  force: Schema.optional(Schema.Boolean),
});
export type GitRemoveWorktreeInput = typeof GitRemoveWorktreeInput.Type;

export const GitCreateBranchInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema,
  publish: Schema.optional(Schema.Boolean),
});
export type GitCreateBranchInput = typeof GitCreateBranchInput.Type;

export const GitCheckoutInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema,
});
export type GitCheckoutInput = typeof GitCheckoutInput.Type;

export const GitStashAndCheckoutInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema,
});
export type GitStashAndCheckoutInput = typeof GitStashAndCheckoutInput.Type;

export const GitStashDropInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitStashDropInput = typeof GitStashDropInput.Type;

export const GitStashInfoInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitStashInfoInput = typeof GitStashInfoInput.Type;

export const GitRemoveIndexLockInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitRemoveIndexLockInput = typeof GitRemoveIndexLockInput.Type;

export const GitInitInput = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
});
export type GitInitInput = typeof GitInitInput.Type;

// RPC Results

const GitStatusPr = Schema.Struct({
  number: PositiveInt,
  title: TrimmedNonEmptyStringSchema,
  url: Schema.String,
  baseBranch: TrimmedNonEmptyStringSchema,
  headBranch: TrimmedNonEmptyStringSchema,
  state: GitStatusPrState,
});

export const GitStatusResult = Schema.Struct({
  branch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  hasWorkingTreeChanges: Schema.Boolean,
  workingTree: Schema.Struct({
    files: Schema.Array(
      Schema.Struct({
        path: TrimmedNonEmptyStringSchema,
        insertions: NonNegativeInt,
        deletions: NonNegativeInt,
      }),
    ),
    insertions: NonNegativeInt,
    deletions: NonNegativeInt,
  }),
  hasUpstream: Schema.Boolean,
  upstreamBranch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  aheadCount: NonNegativeInt,
  behindCount: NonNegativeInt,
  pr: Schema.NullOr(GitStatusPr),
});
export type GitStatusResult = typeof GitStatusResult.Type;

export const GitStatusLocalResult = Schema.Struct({
  branch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  hasWorkingTreeChanges: Schema.Boolean,
  workingTree: GitStatusResult.fields.workingTree,
});
export type GitStatusLocalResult = typeof GitStatusLocalResult.Type;

export const GitStatusRemoteResult = Schema.Struct({
  hasUpstream: Schema.Boolean,
  upstreamBranch: GitStatusResult.fields.upstreamBranch,
  aheadCount: NonNegativeInt,
  behindCount: NonNegativeInt,
  pr: Schema.NullOr(GitStatusPr),
});
export type GitStatusRemoteResult = typeof GitStatusRemoteResult.Type;

export const GitStatusStreamEvent = Schema.Union([
  Schema.TaggedStruct("snapshot", {
    local: GitStatusLocalResult,
    remote: Schema.NullOr(GitStatusRemoteResult),
  }),
  Schema.TaggedStruct("localUpdated", {
    local: GitStatusLocalResult,
  }),
  Schema.TaggedStruct("remoteUpdated", {
    remote: Schema.NullOr(GitStatusRemoteResult),
  }),
]);
export type GitStatusStreamEvent = typeof GitStatusStreamEvent.Type;

export const GitReadWorkingTreeDiffResult = Schema.Struct({
  patch: Schema.String,
});
export type GitReadWorkingTreeDiffResult = typeof GitReadWorkingTreeDiffResult.Type;

export const GitListBranchesResult = Schema.Struct({
  branches: Schema.Array(GitBranch),
  isRepo: Schema.Boolean,
  hasOriginRemote: Schema.Boolean,
});
export type GitListBranchesResult = typeof GitListBranchesResult.Type;

export const GitCreateWorktreeResult = Schema.Struct({
  worktree: GitWorktree,
});
export type GitCreateWorktreeResult = typeof GitCreateWorktreeResult.Type;

export const GitCreateDetachedWorktreeResult = Schema.Struct({
  worktree: GitDetachedWorktree,
});
export type GitCreateDetachedWorktreeResult = typeof GitCreateDetachedWorktreeResult.Type;

export const GitStashInfoResult = Schema.Struct({
  cwd: TrimmedNonEmptyStringSchema,
  branch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  stashRef: TrimmedNonEmptyStringSchema,
  message: TrimmedNonEmptyStringSchema,
  files: Schema.Array(TrimmedNonEmptyStringSchema),
});
export type GitStashInfoResult = typeof GitStashInfoResult.Type;

export const GitResolvePullRequestResult = Schema.Struct({
  pullRequest: GitResolvedPullRequest,
});
export type GitResolvePullRequestResult = typeof GitResolvePullRequestResult.Type;

export const GitPreparePullRequestThreadResult = Schema.Struct({
  pullRequest: GitResolvedPullRequest,
  branch: TrimmedNonEmptyStringSchema,
  worktreePath: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
});
export type GitPreparePullRequestThreadResult = typeof GitPreparePullRequestThreadResult.Type;

export const GitHandoffThreadResult = Schema.Struct({
  targetMode: GitHandoffThreadMode,
  branch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  worktreePath: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  associatedWorktreePath: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  associatedWorktreeBranch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  associatedWorktreeRef: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
  changesTransferred: Schema.Boolean,
  conflictsDetected: Schema.Boolean,
  message: Schema.NullOr(Schema.String),
});
export type GitHandoffThreadResult = typeof GitHandoffThreadResult.Type;

export const GitRunStackedActionResult = Schema.Struct({
  action: GitStackedAction,
  branch: Schema.Struct({
    status: GitBranchStepStatus,
    name: Schema.optional(TrimmedNonEmptyStringSchema),
  }),
  commit: Schema.Struct({
    status: GitCommitStepStatus,
    commitSha: Schema.optional(TrimmedNonEmptyStringSchema),
    subject: Schema.optional(TrimmedNonEmptyStringSchema),
  }),
  push: Schema.Struct({
    status: GitPushStepStatus,
    branch: Schema.optional(TrimmedNonEmptyStringSchema),
    upstreamBranch: Schema.optional(TrimmedNonEmptyStringSchema),
    setUpstream: Schema.optional(Schema.Boolean),
  }),
  pr: Schema.Struct({
    status: GitPrStepStatus,
    url: Schema.optional(Schema.String),
    number: Schema.optional(PositiveInt),
    baseBranch: Schema.optional(TrimmedNonEmptyStringSchema),
    headBranch: Schema.optional(TrimmedNonEmptyStringSchema),
    title: Schema.optional(TrimmedNonEmptyStringSchema),
  }),
});
export type GitRunStackedActionResult = typeof GitRunStackedActionResult.Type;

export const GitPullResult = Schema.Struct({
  status: Schema.Literals(["pulled", "skipped_up_to_date"]),
  branch: TrimmedNonEmptyStringSchema,
  upstreamBranch: TrimmedNonEmptyStringSchema.pipe(Schema.NullOr),
});
export type GitPullResult = typeof GitPullResult.Type;

export const GitSummarizeDiffResult = Schema.Struct({
  summary: TrimmedNonEmptyStringSchema,
});
export type GitSummarizeDiffResult = typeof GitSummarizeDiffResult.Type;

const GitActionProgressBase = Schema.Struct({
  actionId: TrimmedNonEmptyStringSchema,
  cwd: TrimmedNonEmptyStringSchema,
  action: GitStackedAction,
});

const GitActionStartedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("action_started"),
  phases: Schema.Array(GitActionProgressPhase),
});
const GitActionPhaseStartedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("phase_started"),
  phase: GitActionProgressPhase,
  label: TrimmedNonEmptyStringSchema,
});
const GitActionHookStartedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("hook_started"),
  hookName: TrimmedNonEmptyStringSchema,
});
const GitActionHookOutputEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("hook_output"),
  hookName: Schema.NullOr(TrimmedNonEmptyStringSchema),
  stream: GitActionProgressStream,
  text: TrimmedNonEmptyStringSchema,
});
const GitActionHookFinishedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("hook_finished"),
  hookName: TrimmedNonEmptyStringSchema,
  exitCode: Schema.NullOr(Schema.Int),
  durationMs: Schema.NullOr(NonNegativeInt),
});
const GitActionFinishedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("action_finished"),
  result: GitRunStackedActionResult,
});
const GitActionFailedEvent = Schema.Struct({
  ...GitActionProgressBase.fields,
  kind: Schema.Literal("action_failed"),
  phase: Schema.NullOr(GitActionProgressPhase),
  message: TrimmedNonEmptyStringSchema,
});

export const GitActionProgressEvent = Schema.Union([
  GitActionStartedEvent,
  GitActionPhaseStartedEvent,
  GitActionHookStartedEvent,
  GitActionHookOutputEvent,
  GitActionHookFinishedEvent,
  GitActionFinishedEvent,
  GitActionFailedEvent,
]);
export type GitActionProgressEvent = typeof GitActionProgressEvent.Type;
